File: 500/515 - Find Largest Value in Each Tree Row.java
import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> ret = new ArrayList<Integer>();
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        TreeNode curr;
        int len, val;
        
        if (root == null)
            return ret;
        
        queue.add(root);
        while (! queue.isEmpty()) {
            len = queue.size();
            val = Integer.MIN_VALUE;
            
            while (len-- > 0) {
                curr = queue.poll();
                if (curr.left != null)
                    queue.add(curr.left);
                if (curr.right != null)
                    queue.add(curr.right);
                
                val = Math.max(val, curr.val);
            }
            
            ret.add(val);
        }
        
        return ret;
    }
}

File: csharp/515.find-largest-value-in-each-tree-row.cs
/*
 * @lc app=leetcode id=515 lang=csharp
 *
 * [515] Find Largest Value in Each Tree Row
 */

using System.Collections.Generic;

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<int> LargestValues(TreeNode root) {
        var ret = new List<int>();
        if (root == null) {
            return ret;
        }
        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        while (queue.Count > 0) {
            int n = queue.Count;
            int max = int.MinValue;
            while (n-- > 0) {
                var curr = queue.Dequeue();
                if (curr.left != null) {
                    queue.Enqueue(curr.left);
                }
                if (curr.right != null) {
                    queue.Enqueue(curr.right);
                }
                max = Math.Max(max, curr.val);
            }
            ret.Add(max);
        }
        return ret;
    }
}
// @lc code=end

